From 4333456a31c4fe9a8a3c4e52876bf93b9cd991ad Mon Sep 17 00:00:00 2001
From: Steve deRosier <steve.derosier@lairdtech.com>
Date: Wed, 16 Mar 2016 16:32:51 -0700
Subject: [PATCH] laird: wb45: Fix reset command

The 'reset' command wasn't resetting the WB45's processor. Found some fix
code in our old branch, and had to do some minor corrections and fixes to
it. This commit also fixes a few register definitions that weren't quite
right.

Bug: 9111
---
 arch/arm/cpu/arm926ejs/at91/reset.c              | 9 +++++++++
 arch/arm/include/asm/arch-at91/at91sam9_sdramc.h | 2 +-
 arch/arm/include/asm/arch-at91/at91sam9x5.h      | 1 +
 3 files changed, 11 insertions(+), 1 deletion(-)

diff --git a/arch/arm/cpu/arm926ejs/at91/reset.c b/arch/arm/cpu/arm926ejs/at91/reset.c
index e67f47b..4b592bb 100644
--- a/arch/arm/cpu/arm926ejs/at91/reset.c
+++ b/arch/arm/cpu/arm926ejs/at91/reset.c
@@ -11,11 +11,20 @@
 #include <asm/arch/hardware.h>
 #include <asm/arch/at91_rstc.h>
 
+#include <asm/arch/at91sam9_sdramc.h>
+
 /* Reset the cpu by telling the reset controller to do so */
 void reset_cpu(ulong ignored)
 {
 	at91_rstc_t *rstc = (at91_rstc_t *) ATMEL_BASE_RSTC;
 
+#ifdef ATMEL_BASE_SDRAMC
+	void *lpr = (void *) AT91_SDRAMC_LPR;
+
+	/* Ensure that nothing is accessing the SDRAM, as this will inhibit a reset */
+	writel(AT91_SDRAMC_LPCB_DISABLE, lpr);
+#endif
+
 	writel(AT91_RSTC_KEY
 		| AT91_RSTC_CR_PROCRST	/* Processor Reset */
 		| AT91_RSTC_CR_PERRST	/* Peripheral Reset */
diff --git a/arch/arm/include/asm/arch-at91/at91sam9_sdramc.h b/arch/arm/include/asm/arch-at91/at91sam9_sdramc.h
index 5c98cc7..e464b1a 100644
--- a/arch/arm/include/asm/arch-at91/at91sam9_sdramc.h
+++ b/arch/arm/include/asm/arch-at91/at91sam9_sdramc.h
@@ -68,7 +68,7 @@
 #define		AT91_SDRAMC_TRAS	(0xf << 24)		/* Active to Precharge Delay */
 #define		AT91_SDRAMC_TXSR	(0xf << 28)		/* Exit Self Refresh to Active Delay */
 
-#define AT91_SDRAMC_LPR		(ATMEL_BASE_SDRAMC + 0x10)	/* SDRAM Controller Low Power Register */
+#define AT91_SDRAMC_LPR		(ATMEL_BASE_SDRAMC + 0x1C)	/* SDRAM Controller Low Power Register */
 #define		AT91_SDRAMC_LPCB		(3 << 0)	/* Low-power Configurations */
 #define			AT91_SDRAMC_LPCB_DISABLE		0
 #define			AT91_SDRAMC_LPCB_SELF_REFRESH		1
diff --git a/arch/arm/include/asm/arch-at91/at91sam9x5.h b/arch/arm/include/asm/arch-at91/at91sam9x5.h
index d49c184..380974b 100644
--- a/arch/arm/include/asm/arch-at91/at91sam9x5.h
+++ b/arch/arm/include/asm/arch-at91/at91sam9x5.h
@@ -93,6 +93,7 @@
 #define ATMEL_BASE_PMECC	0xffffe000
 #define ATMEL_BASE_PMERRLOC	0xffffe600
 #define ATMEL_BASE_DDRSDRC	0xffffe800
+#define ATMEL_BASE_SDRAMC	0xffffe800
 #define ATMEL_BASE_SMC		0xffffea00
 #define ATMEL_BASE_DMAC0	0xffffec00
 #define ATMEL_BASE_DMAC1	0xffffee00
-- 
1.9.1

